src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -0,0 +1,283 @@
+                }
+
+                auto all_node_ids = ctx.metadata_cache().node_ids();
+                auto unkown_broker_id_it = std::find_if(
+                  partition.replicas->begin(),
+                  partition.replicas->end(),
+                  [all_node_ids](const model::node_id& node_id) {
+                      return std::find(
+                               all_node_ids.begin(),
+                               all_node_ids.end(),
+                               node_id)
+                             == all_node_ids.end();
+                  });
+                if (unkown_broker_id_it != partition.replicas->end()) {
+                    vlog(
+                      klog.error,
+                      "Broker Id is not in the cluster: topic {}, pid {}, "
+                      "broker id "
+                      "{}, live brokers: {}",
+                      topic.name,
+                      partition.partition_index,
+                      *unkown_broker_id_it,
+                      all_node_ids);
+                    resp.data.error_code
+                      = error_code::invalid_replica_assignment;
+                    resp.data.error_message = fmt::format(
+                      "Replica assignment has brokers that are not alive. "
+                      "Replica list: {}, live broker list: {}",
+                      *partition.replicas,
+                      all_node_ids);
+                    co_return co_await ctx.respond(std::move(resp));
+                }
+
+                auto assignment_exists
+                  = co_await ctx.topics_frontend().replica_set_exists(
+                    ntp, *partition.replicas);
+                if (assignment_exists) {
+                    // The requested reassignment already exists, so skip this
+                    // one
+                    vlog(
+                      klog.info,
+                      "Request to change ntp {} replica set to {}, no change",
+                      ntp,
+                      *partition.replicas);
+                    continue;
+                }
+
+                // Convert each node id to a broker shard because
+                // topics_frontend.move_partition_replicas expects
+                // vector<broker_shard>
+                std::vector<model::broker_shard> replicas;
+                std::for_each(
+                  partition.replicas->cbegin(),
+                  partition.replicas->cend(),
+                  [&replicas](const model::node_id& node_id) {
+                      replicas.push_back(
+                        model::broker_shard{.node_id = node_id, .shard = 0});
+                  });
+
+                vlog(
+                  klog.debug,
+                  "Request to reassign partitions: ntp {}, replicas {}",
+                  ntp,
+                  replicas);
+
+                auto errc
+                  = co_await ctx.topics_frontend().move_partition_replicas(
```

Comment:
If there is no strange requests like attempts to move a single partition multiple times within one request we could execute these in parallel.

If those requests do exist like (move [partition1-> node2] then [partition1->node3]) we could just omit the move from p1-n2 and just go straight to p1->n3, then we're sure we can execute this in parallel without worrying about request ordering.

Just a thought, maybe theres no additional benefit to this though @dotnwat ?

Reply:
This sounds like an optimization that could wait until after the initial feature is merged and tests are made.

Reply:
Let's do them serially. The Kafka layer should not be managing that sort of concurrency / scheduling. If we want to speed it up by running movements in parallel, then I think that the controller should provide an interface where we can hand off a bunch of work.